core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -924,49 +939,74 @@ private[deploy] class Master(
    * Return the UI if successful, else None
    */
   private[master] def rebuildSparkUI(app: ApplicationInfo): Option[SparkUI] = {
+    val futureUI = asyncRebuildSparkUI(app)
+
+    // Wait until the UI is rebuilt or cancelled, log status occasionally for apps with many events
+    var waitSec = 2
+    do {
+      try {
+        val ui = Await.result(futureUI, Duration(waitSec, TimeUnit.SECONDS))
+        return Some(ui)
+      } catch {
+        case te: TimeoutException =>
+          waitSec *= 2
+          logInfo(s"Application UI is rebuilding, will continue to wait $waitSec seconds")
+        case NonFatal(_) | _: InterruptedException =>
+          waitSec = 0
+      }
+    } while (waitSec > 0)
```

Comment:
This feature could be useful, but it's technically not SPARK-12062 and only used in the REST API now. Especially since we plan to remove rebuilding this on the Master altogether in Spark 2.0 I'd just await on the future for simplicity.